Documentation and changes to JournalProxy (PR #18876)

talp-traiana opened the pull request:
- Add note in Shared LevelDB saying Journal Proxy should be used instead.
- Still refers to the plugin as JournalProxy (not PersistencePluginProxy).
- Documented the ability to set target via configuration (configuration key was left TBD).

andreaTP committed:
=act split FSM java api in a separate file.

andreaTP committed:
=act split Props java api implementation in separate files.

andreaTP committed:
=act split ReflectiveDynamicAccess implementation in separate file.

andreaTP committed:
=act Scheduler implementation separated from interface.

Commit:
+rem #18353: Prune reliable deliver actors
(cherry picked from commit 6643f56)

Commit:
+clu #17695 add a way to inspect the current sharding state

Two new message pairs:
`GetShardRegionState`/`CurrentShardRegionState` allows for querying a region for it's current shards and the current `EntityIds` of it
`GetClusterShardingStats`/`ClusterShardingStats` allows for querying the entire cluster for a summary of
the number of entitites alive in each region and shard.

andreaTP committed:
=act switch to java std lib ThreadLocalRandom

Commit:
+act #18575 Publish MemberJoined

patriknw committed:
=cdd #18768 Cache serialization of read and write msg

The WriteAggregator and ReadAggregator typically send
the same message to several replicas and by caching the serialized bytes
we avoid doing the same thing for each message

and add a test for WriteAggregator

patriknw committed:
=clu #18554 Make oldest assignment deterministic when joining

* the reported issue is fixed by the immediate leaderActions
  (moving to Up)  when joining the first node to itself
* the other changes are precautions just in case

patriknw committed:
=act #18613 Optimize RepointableActorRef for sys msg

* Before the cell is started the RepointableActorRef queue the messages
  (both system and non-system messages) into a queue. System messages
  are inserted right before the last non-system message by iterating
  the queue, i.e. becomes slow when there are many system messages already
  enqueued.
* This is a problem for pools with many routees.
* The fix is to keep system messages in a separate queue

JMH benchmark before:

    [info] Benchmark                                 (size)  Mode  Cnt       Score      Error  Units
    [info] RouterPoolCreationBenchmark.testCreation    1000    ss  300   13204.048 ± 3081.576  us/op
    [info] RouterPoolCreationBenchmark.testCreation    2000    ss  300   41939.524 ± 6178.087  us/op
    [info] RouterPoolCreationBenchmark.testCreation    3000    ss  300   70752.881 ± 4344.992  us/op
    [info] RouterPoolCreationBenchmark.testCreation    4000    ss  300  120620.885 ± 3296.342  us/op

JMH benchmark after:

    [info] Benchmark                                 (size)  Mode  Cnt      Score      Error  Units
    [info] RouterPoolCreationBenchmark.testCreation    1000    ss  300   7738.721 ± 1806.297  us/op
    [info] RouterPoolCreationBenchmark.testCreation    2000    ss  300  15497.588 ± 4532.852  us/op
    [info] RouterPoolCreationBenchmark.testCreation    3000    ss  300  28704.005 ± 6322.458  us/op
    [info] RouterPoolCreationBenchmark.testCreation    4000    ss  300  37783.516 ± 6778.437  us/op

fully drain sysmsgQueue when swap cell

patriknw committed:
=act #16522 Optimize Stash creation

* cache the stash capacity config value

BEFORE

    [info] Result "testCustom":
    [info]   27.607 ±(99.9%) 7.924 us/op [Average]
    [info]   (min, avg, max) = (16.928, 27.607, 760217.600), stdev = 1369.808
    [info]   CI (99.9%): [19.682, 35.531] (assumes normal distribution)
    [info]   Samples, N = 323534
    [info]         mean =     27.607 ±(99.9%) 7.924 us/op
    [info]          min =     16.928 us/op
    [info]   p( 0.0000) =     16.928 us/op
    [info]   p(50.0000) =     21.824 us/op
    [info]   p(90.0000) =     31.680 us/op
    [info]   p(95.0000) =     38.592 us/op
    [info]   p(99.0000) =     57.920 us/op
    [info]   p(99.9000) =    102.144 us/op
    [info]   p(99.9900) =    371.471 us/op
    [info]   p(99.9990) =  85461.364 us/op
    [info]   p(99.9999) = 760217.600 us/op
    [info]          max = 760217.600 us/op

    [info] Result "testDefault":
    [info]   26.465 ±(99.9%) 15.980 us/op [Average]
    [info]   (min, avg, max) = (14.608, 26.465, 1461714.944), stdev = 2900.094
    [info]   CI (99.9%): [10.485, 42.444] (assumes normal distribution)
    [info]   Samples, N = 356626
    [info]         mean =     26.465 ±(99.9%) 15.980 us/op
    [info]          min =     14.608 us/op
    [info]   p( 0.0000) =     14.608 us/op
    [info]   p(50.0000) =     18.304 us/op
    [info]   p(90.0000) =     22.816 us/op
    [info]   p(95.0000) =     26.976 us/op
    [info]   p(99.0000) =     41.647 us/op
    [info]   p(99.9000) =     71.808 us/op
    [info]   p(99.9900) =    139.696 us/op
    [info]   p(99.9990) =  33048.112 us/op
    [info]   p(99.9999) = 1461714.944 us/op
    [info]          max = 1461714.944 us/op

    [info] Benchmark                             Mode     Cnt   Score    Error  Units
    [info] StashCreationBenchmark.testCustom   sample  323534  27.607 ±  7.924  us/op
    [info] StashCreationBenchmark.testDefault  sample  356626  26.465 ± 15.980  us/op

AFTER:

    [info] Result "testCustom":
    [info]   22.710 ±(99.9%) 1.384 us/op [Average]
    [info]   (min, avg, max) = (15.168, 22.710, 110886.912), stdev = 233.744
    [info]   CI (99.9%): [21.326, 24.094] (assumes normal distribution)
    [info]   Samples, N = 308896
    [info]         mean =     22.710 ±(99.9%) 1.384 us/op
    [info]          min =     15.168 us/op
    [info]   p( 0.0000) =     15.168 us/op
    [info]   p(50.0000) =     19.776 us/op
    [info]   p(90.0000) =     28.608 us/op
    [info]   p(95.0000) =     33.344 us/op
    [info]   p(99.0000) =     49.216 us/op
    [info]   p(99.9000) =     90.253 us/op
    [info]   p(99.9900) =    594.334 us/op
    [info]   p(99.9990) =   8350.329 us/op
    [info]   p(99.9999) = 110886.912 us/op
    [info]          max = 110886.912 us/op

    [info] Result "testDefault":
    [info]   19.099 ±(99.9%) 0.424 us/op [Average]
    [info]   (min, avg, max) = (13.120, 19.099, 45940.736), stdev = 77.524
    [info]   CI (99.9%): [18.675, 19.523] (assumes normal distribution)
    [info]   Samples, N = 361741
    [info]         mean =     19.099 ±(99.9%) 0.424 us/op
    [info]          min =     13.120 us/op
    [info]   p( 0.0000) =     13.120 us/op
    [info]   p(50.0000) =     16.368 us/op
    [info]   p(90.0000) =     24.250 us/op
    [info]   p(95.0000) =     31.232 us/op
    [info]   p(99.0000) =     55.680 us/op
    [info]   p(99.9000) =    121.249 us/op
    [info]   p(99.9900) =    402.386 us/op
    [info]   p(99.9990) =   1767.264 us/op
    [info]   p(99.9999) =  45940.736 us/op
    [info]          max =  45940.736 us/op
    [info]
    [info]
    [info] # Run complete. Total time: 00:02:01
    [info]
    [info] Benchmark                             Mode     Cnt   Score   Error  Units
    [info] StashCreationBenchmark.testCustom   sample  308896  22.710 ± 1.384  us/op
    [info] StashCreationBenchmark.testDefault  sample  361741  19.099 ± 0.424  us/op

ataraxer committed:
+act #15819 Create ask() variant that can be used without sender()

ataraxer committed:
+act #15819 Add PromiseRef and FutureRef as an alternative to ask pattern

henrymai committed:
Add TransparentExponentialBackoffSupervisor

Commit:
=cls #18762 fix graceful shutdown of empty region

Commit:
consider akka.test.timefactor when using akka.test.filter-leeway

Commit:
+cam #11214 add attachments to CamelMessage

add support for attachments in CamelMessage, taken from JCamelMessage

add a extension class CamelMessageWithAttachments base on CamelMessage that contains attachments based on JCamelMessage

add newline CamelMessageTest

replace for-cycle attachment adding with one putAll

remove CamelMessageWithAttachments and update CamelMessage to contain attachments

fix the binary compatibility issues on CamelMessage

add documentation for new methods in CamelMessage, add return type for method withBody

patriknw committed:
=clu replace Set -- with diff and ++ with union

* better performance according to
  https://docs.google.com/presentation/d/1Qjryxoe-fYEM8ZPhM-98LKfbhnRcn5eAEMNlVVnixsA/pub

patriknw committed:
Merge pull request #18769 from akka/wip-ddata-perf-many-nodes-patriknw

=cdd #18768 Cache serialization of read and write msg

patriknw committed:
Merge pull request #18746 from akka/wip-18554-singleton-startup-patriknw

=clu #18554 Make oldest assignment deterministic when joining

patriknw committed:
+per #18839 Add better journal proxy

* JournalProxy works on the journal and snapshot store
  message protocol level and is therefore much simpler
  than the previous AsyncWriteProxy approach
* JournalProxy will work out of the box with any journal,
  I have tried with Cassandra
* should in the end phase out AsyncWriteProxy and
  SharedLeveldbJournal, we need deprecation phase
* rewrote the InmemJournal to remove the dependency to
  AsyncWriteProxy, which I can't see why it is needed
  for the InmemJournal

rkuhn committed:
Merge pull request #18776 from henrymai/master

Add TransparentExponentialBackoffSupervisor

rkuhn committed:
Merge pull request #18875 from ktoso/wip-akka.js-cherries-ktoso

Akka.js cherries to master

ktoso committed:
Merge pull request #18755 from akka/wip-16522-optimize-stash-config-patriknw

=act #16522 Optimize Stash creation

Commit:
=doc #18635 Added section about `ActorSystem.shutdown` being deprecated

kukido committed:
=doc Fixes for genjavadoc

Commit:
Merge pull request #18543 from johanandren/wip-17695-api-for-sharding-state-retrieval-johanandren

+clu #17695 add a way to inspect the current sharding state

Commit:
Merge pull request #18729 from hseeberger/hseeberger-18575-publish-member-joined

Publish MemberJoined

Commit:
Merge pull request #18810 from swachter/master

consider akka.test.timefactor when using akka.test.filter-leeway

Commit:
=act #18607 Test failing on timeout not happening

Changed the future to never complete to be sure there is no racing

ktoso committed:
Merge pull request #18909 from kukido/kukido-javadoc-fixes

=doc Fixes for genjavadoc

ktoso committed:
Merge pull request #18911 from ktoso/wip-rebased-18805-ktoso

[rebased] =cls #18762 fix graceful shutdown of empty region

ktoso committed:
Merge pull request #18907 from johanandren/wip-18635-migration-guide-terminate-johanandren

=doc #18635 Added section about `ActorSystem.shutdown` being deprecated

kukido committed:
=doc #18899 code review follow-up, fixed typo

Commit:
Merge pull request #18912 from johanandren/wip-18607-no-timeout-ex-waiting-for-future-johanandren

=act #18607 Test failing on timeout not happening

Commit:
Merge pull request #18754 from akka/wip-18613-optimize-repointable-patriknw

=act #18613 Optimize RepointableActorRef for only sys msg

ktoso committed:
Merge pull request #18915 from kukido/kukido-18899-pr-follow-up

=doc #18899 code review follow-up, fixed typo

szymon-rd committed:
=doc Fixed typo

ktoso committed:
Merge pull request #18932 from Jaca777/jaca777-4761866

=doc Fixed typo

szymon-rd committed:
=pro Fix unmaintained scalariform github repo link.

ktoso committed:
Merge pull request #18933 from Jaca777/jaca777-4761866

=pro Fix unmaintained scalariform github repo link.

Commit:
Merge pull request #18920 from drewhk/wip-18353-fwd-drewhk

+rem #18353: Prune reliable deliver actors (fwd)

patriknw committed:
+act #18487 Support custom supervision strategy in BackoffSupervisor

* and forward messages from child to parent
  (needed for sharding passivation)

talpr committed:
Add documentation for JournalProxy:
* Add note in Shared LevelDB saying Journal Proxy should be used instead.
* Still refers to the plugin as JournalProxy (not PersistencePluginProxy).
* Documented the ability to set target via configuration (configuration key was left TBD).

patriknw commented on akka-docs/rst/java/lambda-persistence.rst:
```diff
@@ -892,6 +892,10 @@ backup node.
 
+.. note::
+
+  This plugin has been supplanted by :ref:`Journal Proxy <journal-proxy>`.
```
We should also start a migration guide migration-guide-2.4.x-2.5.x.rst, use https://github.com/akka/akka/blob/master/akka-docs/rst/project/migration-guide-2.3.x-2.4.x.rst as template. Nothing fancy, just mentioning what you wrote here in the note.

patriknw commented on akka-docs/rst/java/lambda-persistence.rst:
```diff
@@ -932,6 +936,37 @@ directory. This can be changed by configuration where the specified path can be
+different nodes). This, for example, allows persistent actors to failover to a backup node and continue using the
+shared journal instance from the backup node. The journal proxy works by forwarding all the journal/snapshot store
+messages to a single, shared, persistence plugin instance, and therefor supports any use case supported by the proxied
+plugin (for example, :ref:`Persistence Query <persistence-query-scala>`).
```
is this the right link? I don't think query plugins can be proxied.

talp-traiana replied:
I've tested the LevelDB persistence query with the journal and it works (I also checked the code, and all queries forward messages to the journal).
I'm not sure that's actually what was intended for PQ, but I would argue that if it doesn't communicate with the persistence layer via the plugin then there's a pretty strong dependency on the exact data format the journal uses.

patriknw replied:
ah, that's just because the leveldb query is using the write journal to serve the queries, I don't think that will be a normal solution, so we should not promise anything about that in the docs

patriknw commented on akka-docs/rst/java/persistence.rst:
```diff
@@ -918,6 +922,37 @@ directory. This can be changed by configuration where the specified path can be
 Note that it is not mandatory to specify a snapshot store plugin. If you don't use snapshots
 you don't have to configure it.
 
+.. _journal-proxy:
```
I think those must be unique, so include the `java` as we have done in other places

patriknw commented:
OK TO TEST

patriknw commented:
looking great, thanks

rkuhn committed:
Merge pull request #18815 from visox/master

Add attachments in CamelMessage

rkuhn committed:
Merge pull request #18243 from ataraxer/ataraxer-15819

+act #15819 Create ask() variant that can be used without sender(); Add PromiseRef as an alternative to ask pattern

patriknw committed:
=cls #18722 fix DDataShardCoordinator init

* the become logic was wrong when watchStateActors triggers an immediate
  state update

Commit:
=tes #18312 Increased the timeout of the test actor and dilated it

talpr committed:
* Fixed Java and Java8 tag names
* Picked configuration key for setting target via configuration

Commit:
+act #18913 moved PrettyDuration to akka-actor

andreaTP committed:
=act removed reference to java.reflect.

andreaTP committed:
=act move lookupAndSetField to Reflect helper.

talpr committed:
Docs:
* Remove reference to Persistence Query
* Add migration guide

talpr committed:
Set target plugin address via configuration.

talpr committed:
Rename JournalProxy to PersistencePluginProxy.

talpr committed:
Log the timeout duration

talpr committed:
Switch to ask instead of forward so no messages are lost

talp-traiana commented:
TODO:
1. Handle future timeouts properly
2. Only do initialization-by-config if startTarget is false

ktoso committed:
Merge pull request #18980 from andreaTP/indirectFix

=act removed reference to java.reflect.

talpr committed:
* Handle ask timeouts for protocol messages
* Don't try to set the address on the node that holds the target

Commit:
Merge pull request #18931 from lolski/lolski-issue-18913

moving PrettyDuration and PrettyDurationSpec to akka.util

jypma committed:
!act #18996 PFBuilder matchAny should take declared input type I

Currently PFBuilder.matchAny takes a lambda with Object as input argument.
This loses type information for partial functions that are not akka
receive functions, e.g. exception handlers (PF<Exception,RouteResult).
With this change, the fallback handler still has the guaranteed type I
visible on its lambda.

talp-traiana commented:
@patriknw 
I did some more testing, and ran into an issue I thought you might have some ideas about.
It appears that many of the journal messages contain the ActorRef to respond to as part of the message. This causes the ask futures to time-out since they're not the ones that get the response.

There are 3 ways around this that I can think of, but I don't like any of them:
1. Change AsyncWriteJournal and the journal protocol messages to always reply to sender. Don't like it, because I assume there's a good reason it was implemented this way in the first place.
2. Add a mechanism that keeps tracks of requests and responses in the proxy. This seems like unnecessary complication of the proxy code, for something that seems pretty simple.
3. Extend AskSupport with an ask that takes an (ActorRef => Any), and inject ask's PromiseActorRef into the message. The downside to this is that it's messing with something very general (the ask mechanism) for a very specific use-case. It's also possible to just duplicate the relevant ask code, but then we get code duplication...

What do you think?

patriknw committed:
Merge pull request #18960 from akka/wip-18722-sharding-ddata-init-patriknw

=cls #18722 fix DDataShardCoordinator init

patriknw commented:
@talp-traiana 

> What should we do regarding "regular" messages that are not part of the protocol (ActorIdentity, etc.)?

just forward them, there is not anything else we can do

> sender

I think the reason is that we need to keep track of two things. 1) the PersistentActor ref to send the replies to, 2) the original sender of the application, i.e. the sender of the command in the PersistentActor

Can you take a look at https://github.com/akka/akka/pull/18243 and see if we can use that? It was recently merged to master.

patriknw committed:
Merge pull request #18944 from akka/wip-18487-BackofSupervisor-decider-patriknw

+act #18487 Support custom supervision strategy in BackoffSupervisor

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/PersistencePluginProxy.scala:
```diff
@@ -84,6 +68,17 @@ final class JournalProxy(config: Config) extends Actor with Stash with ActorLogg
       context.become(active(target, targetAtThisNode = true))
     } else {
+      val targetAddressKey = s"target-${pluginType.qualifier}-address"
+      if (config.hasPath(targetAddressKey)) {
```
We normally define the property in the `reference.conf` anyway, with a placeholder value, such as empty string `""`.
In that way we can document the property and it is more visible that it exists.

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/PersistencePluginProxy.scala:
```diff
@@ -136,8 +131,57 @@ final class JournalProxy(config: Config) extends Actor with Stash with ActorLogg
+
+    case msg @ WriteMessages(messages, _, actorInstanceId) ⇒
+      implicit val ec = context.system.dispatcher
+      val sender = context.sender()
```
use another name than `sender`, e.g. `replyTo`

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/PersistencePluginProxy.scala:
```diff
@@ -136,8 +131,57 @@ final class JournalProxy(config: Config) extends Actor with Stash with ActorLogg
+      implicit val ec = context.system.dispatcher
+      val sender = context.sender()
+      val f = ask(targetJournal, msg)(1.second)
+      f onComplete {
```
use `recover` + `pipeTo` as in the other cases

talp-traiana replied:
From what I understand, pipeTo only returns 1 message whereas when this future fails we may need to send the actor several messages. I'm open to suggestions on how to make it cleaner, though.

patriknw replied:
ah, I missed that, you are good with `onComplete` then

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/PersistencePluginProxy.scala:
```diff
@@ -136,8 +131,57 @@ final class JournalProxy(config: Config) extends Actor with Stash with ActorLogg
+      val f = ask(targetJournal, msg)(2.seconds)
+      f recover { case t ⇒ DeleteSnapshotsFailure(criteria, t) } pipeTo sender
+
+    case msg ⇒ // is this necessary? what other messages can we get?
```
for custom plugin defined messages, let's just forward them (we can't know if there will be a reply)

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/PersistencePluginProxy.scala:
```diff
@@ -136,8 +131,57 @@ final class JournalProxy(config: Config) extends Actor with Stash with ActorLogg
+    case msg @ DeleteMessagesTo(_, toSeqNr, _) ⇒
+      implicit val ec = context.system.dispatcher
+      val sender = context.sender()
+      val f = ask(targetJournal, msg)(2.seconds)
```
define the timeout in a config property?

patriknw replied:
and use a longer default, e.g. 10 seconds (some journals are slow when starting up in testing mode, e.g. embedded cassandra)

patriknw committed:
Merge pull request #18998 from jypma/pfbuilder_api_master

PFBuilder matchAny should take declared input type I

ktoso committed:
Merge pull request #18963 from johanandren/wip-18312-failed-cluster-metrics-test-johanandren

=tes #18312 Increased the timeout of the test actor and dilated it

ktoso committed:
Merge pull request #18981 from andreaTP/setField

=act move lookupAndSetField to Reflect helper.

talp-traiana commented:
Hmm, I rebased on master to get #18243, and made a mess of this PR. Should I close this and open a new one?

patriknw commented:
do whatever is easiest for you

talp-traiana commented:
Replaced by #19008
